perf(treesitter): partial injection and incremental invalidation

tomtomjhj opened the pull request:
The two main bottlenecks in editing big files with many injections are (1) parsing and (2) processing the injection queries. This PR resolves (2) by applying injection query to only visible regions.

<!--
Status:
* Seems to work, but probably needs more testing.
* There are some TODOs regarding changes that may affect API users (maybe related to failing test cases).
-->

Benchmark:
1. download <https://raw.githubusercontent.com/torvalds/linux/master/drivers/gpu/drm/amd/include/asic_reg/dcn/dcn_3_2_0_sh_mask.h>
1. setup nvim-treesitter for `cpp` (`.h` file is recognized as `cpp` filetype by default)
1. `:let g:__ts_debug = 1`
1. `:e dcn_3_2_0_sh_mask.h`, `:lua vim.treesitter.start()`
1. go to the last comment in the file (`G?//<CR>`), then append a space (`A<Space><Esc>`)
1. go to the next line (`j`) and append 0 (`A0<Esc>`)
1. see log at `~/.local/state/nvim/treesitter.log`

<details>
<summary>before</summary>

```
1:cpp:(nvim) parse:450: (#regions=1)  { parse_time = 2923.453483, query_time = 0, regions_parsed = 1 }
1:cpp:(nvim) parse:450: (#regions=1)  { parse_time = 0, query_time = 0, regions_parsed = 0 }
1:cpp:(nvim) parse:450: (#regions=1)  { parse_time = 0, query_time = 0, regions_parsed = 0 }
1:comment:(nvim) new:145: (#regions=1)  START
1:cpp:(nvim) new:145: (#regions=1)  START
1:cpp:(nvim) parse:450: (#regions=1)  { parse_time = 0, query_time = 1004.06346, range = { 0, 56 }, regions_parsed = 0 }
1:cpp:(nvim) parse:450: (#regions=194402)  { parse_time = 1.506477, query_time = 0.17726, range = { 0, 56 }, regions_parsed = 25 }
1:comment:(nvim) parse:450: (#regions=27859)  { parse_time = 1.215564, query_time = 0.001373, range = { 0, 56 }, regions_parsed = 5 }
1:cpp:(nvim) parse:450: (#regions=1)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:450: (#regions=194402)  { parse_time = 2.845539, query_time = 0.450294, range = { 222837, 222893 }, regions_parsed = 50 }
1:comment:(nvim) parse:450: (#regions=27859)  { parse_time = 0.266351, query_time = 0.001112, range = { 222837, 222893 }, regions_parsed = 4 }
1:cpp:(nvim) parse:450: (#regions=1)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:450: (#regions=194402)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
1:comment:(nvim) parse:450: (#regions=27859)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:450: (#regions=1)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:450: (#regions=194402)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
1:comment:(nvim) parse:450: (#regions=27859)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:450: (#regions=1)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:450: (#regions=194402)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
1:comment:(nvim) parse:450: (#regions=27859)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:450: (#regions=1)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:450: (#regions=194402)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
1:comment:(nvim) parse:450: (#regions=27859)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) _on_bytes:984: (#regions=1)  on_bytes 1 4 222884 62 23943921 0 0 0 0 1 1
1:cpp:(nvim) _iter_regions:571: (#regions=1)  was valid true
1:cpp:(nvim) _iter_regions:581: (#regions=1)  invalidating region 1 []
1:comment:(nvim) _iter_regions:581: (#regions=9)  invalidating region 27859 [222884:0-222884:63]
1:cpp:(nvim) parse:450: (#regions=1)  { parse_time = 366.826656, query_time = 1008.172019, range = { 222837, 222893 }, regions_parsed = 1 }
1:cpp:(nvim) parse:450: (#regions=194402)  { parse_time = 3.726473, query_time = 0.987651, range = { 222837, 222893 }, regions_parsed = 50 }
1:comment:(nvim) parse:450: (#regions=27859)  { parse_time = 0.193249, query_time = 0.006221, range = { 222837, 222893 }, regions_parsed = 4 }
1:cpp:(nvim) parse:450: (#regions=1)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:450: (#regions=194402)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
1:comment:(nvim) parse:450: (#regions=27859)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:450: (#regions=1)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:450: (#regions=194402)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
1:comment:(nvim) parse:450: (#regions=27859)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:450: (#regions=1)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:450: (#regions=194402)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
1:comment:(nvim) parse:450: (#regions=27859)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) _on_bytes:984: (#regions=1)  on_bytes 1 5 222885 113 23944036 0 0 0 0 1 1
1:cpp:(nvim) _iter_regions:571: (#regions=1)  was valid true
1:cpp:(nvim) _iter_regions:581: (#regions=1)  invalidating region 1 []
1:cpp:(nvim) _iter_regions:581: (#regions=50)  invalidating region 194397 [222885:110-222885:114]
1:cpp:(nvim) parse:450: (#regions=1)  { parse_time = 391.276224, query_time = 1176.385789, range = { 222837, 222893 }, regions_parsed = 1 }
1:cpp:(nvim) parse:450: (#regions=194402)  { parse_time = 3.136393, query_time = 0.262249, range = { 222837, 222893 }, regions_parsed = 50 }
1:comment:(nvim) parse:450: (#regions=27859)  { parse_time = 0.135404, query_time = 0.000935, range = { 222837, 222893 }, regions_parsed = 4 }
1:cpp:(nvim) parse:450: (#regions=1)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:450: (#regions=194402)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
1:comment:(nvim) parse:450: (#regions=27859)  { parse_time = 0, query_time = 0, range = { 222837, 222893 }, regions_parsed = 0 }
```
</details>

<details>
<summary>after</summary>

```
1:cpp:(nvim) parse:484: (#regions=1)  { parse_time = 2782.294202, query_time = 0, regions_parsed = 1 }
1:cpp:(nvim) parse:484: (#regions=1)  { parse_time = 0, query_time = 0, regions_parsed = 0 }
1:cpp:(nvim) parse:484: (#regions=1)  { parse_time = 0, query_time = 0, regions_parsed = 0 }
1:comment:(nvim) new:162: (#regions=1)  START
1:cpp:(nvim) new:162: (#regions=1)  START
1:cpp:(nvim) parse:484: (#regions=1)  { parse_time = 0, query_time = 0.38573, range = { 0, 56 }, regions_parsed = 0 }
1:comment:(nvim) parse:484: (#regions=5)  { parse_time = 0.725387, query_time = 0.000461, range = { 0, 56 }, regions_parsed = 5 }
1:cpp:(nvim) parse:484: (#regions=25)  { parse_time = 0.898121, query_time = 0.061065, range = { 0, 56 }, regions_parsed = 25 }
1:cpp:(nvim) parse:484: (#regions=1)  { parse_time = 0, query_time = 1.598029, range = { 222837, 222893 }, regions_parsed = 0 }
1:comment:(nvim) parse:484: (#regions=9)  { parse_time = 0.354243, query_time = 0.001245, range = { 222837, 222893 }, regions_parsed = 4 }
1:cpp:(nvim) parse:484: (#regions=75)  { parse_time = 10.392799, query_time = 0.823553, range = { 222837, 222893 }, regions_parsed = 50 }
1:cpp:(nvim) parse:484: (#regions=1)  { parse_time = 0, query_time = 0.96193, range = { 222837, 222893 }, regions_parsed = 0 }
1:comment:(nvim) parse:484: (#regions=4)  { parse_time = 0, query_time = 0.003102, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:484: (#regions=50)  { parse_time = 0, query_time = 0.720424, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:484: (#regions=1)  { parse_time = 0, query_time = 1.32763, range = { 222837, 222893 }, regions_parsed = 0 }
1:comment:(nvim) parse:484: (#regions=4)  { parse_time = 0, query_time = 0.0019, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:484: (#regions=50)  { parse_time = 0, query_time = 0.661419, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:484: (#regions=1)  { parse_time = 0, query_time = 1.424348, range = { 222837, 222893 }, regions_parsed = 0 }
1:comment:(nvim) parse:484: (#regions=4)  { parse_time = 0, query_time = 0.001834, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:484: (#regions=50)  { parse_time = 0, query_time = 0.769533, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:484: (#regions=1)  { parse_time = 0, query_time = 1.261363, range = { 222837, 222893 }, regions_parsed = 0 }
1:comment:(nvim) parse:484: (#regions=4)  { parse_time = 0, query_time = 0.001738, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:484: (#regions=50)  { parse_time = 0, query_time = 0.596901, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) _on_bytes:1144: (#regions=1)  on_bytes 1 4 222884 62 23943921 0 0 0 0 1 1
1:cpp:(nvim) _iter_regions:606: (#regions=1)  was valid true
1:cpp:(nvim) _iter_regions:616: (#regions=1)  invalidating region 1 []
1:comment:(nvim) _iter_regions:606: (#regions=4)  was valid true
1:comment:(nvim) _iter_regions:616: (#regions=4)  invalidating region 9 [222884:0-222884:63]
1:cpp:(nvim) _iter_regions:606: (#regions=50)  was valid true
1:cpp:(nvim) parse:484: (#regions=1)  { parse_time = 399.244173, query_time = 0.407459, range = { 222837, 222893 }, regions_parsed = 1 }
1:comment:(nvim) parse:484: (#regions=4)  { changes = { { 222884, 62, 23943921, 222884, 63, 23943922 } }, parse_time = 0.046848, query_time = 0.00045, range = { 222837, 222893 }, regions_parsed = 1 }
1:cpp:(nvim) parse:484: (#regions=50)  { parse_time = 0, query_time = 0.197181, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:484: (#regions=1)  { parse_time = 0, query_time = 0.976214, range = { 222837, 222893 }, regions_parsed = 0 }
1:comment:(nvim) parse:484: (#regions=4)  { parse_time = 0, query_time = 0.002079, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:484: (#regions=50)  { parse_time = 0, query_time = 0.655016, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:484: (#regions=1)  { parse_time = 0, query_time = 1.166834, range = { 222837, 222893 }, regions_parsed = 0 }
1:comment:(nvim) parse:484: (#regions=4)  { parse_time = 0, query_time = 0.00189, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:484: (#regions=50)  { parse_time = 0, query_time = 0.893842, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:484: (#regions=1)  { parse_time = 0, query_time = 1.057452, range = { 222837, 222893 }, regions_parsed = 0 }
1:comment:(nvim) parse:484: (#regions=4)  { parse_time = 0, query_time = 0.001859, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:484: (#regions=50)  { parse_time = 0, query_time = 0.650981, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) _on_bytes:1144: (#regions=1)  on_bytes 1 5 222885 113 23944036 0 0 0 0 1 1
1:cpp:(nvim) _iter_regions:606: (#regions=1)  was valid true
1:cpp:(nvim) _iter_regions:616: (#regions=1)  invalidating region 1 []
1:comment:(nvim) _iter_regions:606: (#regions=4)  was valid true
1:cpp:(nvim) _iter_regions:606: (#regions=50)  was valid true
1:cpp:(nvim) _iter_regions:616: (#regions=50)  invalidating region 70 [222885:110-222885:114]
1:cpp:(nvim) parse:484: (#regions=1)  { parse_time = 376.665228, query_time = 0.288334, range = { 222837, 222893 }, regions_parsed = 1 }
1:comment:(nvim) parse:484: (#regions=4)  { parse_time = 0, query_time = 0.000528, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:484: (#regions=50)  { parse_time = 0.076109, query_time = 0.217351, range = { 222837, 222893 }, regions_parsed = 1 }
1:cpp:(nvim) parse:484: (#regions=1)  { parse_time = 0, query_time = 1.539355, range = { 222837, 222893 }, regions_parsed = 0 }
1:comment:(nvim) parse:484: (#regions=4)  { parse_time = 0, query_time = 0.002121, range = { 222837, 222893 }, regions_parsed = 0 }
1:cpp:(nvim) parse:484: (#regions=50)  { parse_time = 0, query_time = 0.945593, range = { 222837, 222893 }, regions_parsed = 0 }
```
</details>

Evaluation
* Major benefit: The query time becomes negligible. Now the only bottleneck is the parsing of the root cpp parser.
* Minor additional benefit: less reparsing of injections
* Minor downside: It runs the injection query more frequently, for example when scrolling and when multiple windows show the same buffer (not shown in the above logs). But the cost is negligible.

tomtomjhj commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -822,30 +965,47 @@ end
+  local range_start_line, range_end_line ---@type integer, integer
+  if range ~= true then
+    range_start_line, _, range_end_line, _ = Range.unpack4(range)
+  end
```
There is a guard on `range == false` and `range == nil` cases at the top of this function: don't run injection query at all.

tomtomjhj commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -822,30 +965,47 @@ end
+    -- match each pattern separately.
+    if not self._has_combined_injection and range ~= true then
+      start_line = math.max(start_line, range_start_line)
+      end_line = math.min(end_line, range_end_line)
```
In this branch `range` must have type `Range`, so `range_start_line`, `range_end_line` must've been set.

ribru17 replied:
If this is the case then I think `range` should not be defined to allow a `boolean` type at all? Rather it should be typed only as `Range`, no?

It is confusing to see it span a boolean type as well, and especially since passing in `true` leads to nil access in the `math.max`

tomtomjhj commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -131,13 +134,27 @@ function LanguageTree.new(source, lang, opts, parent_lang)
+      if
+        vim.tbl_contains(preds, function(pred)
+          return vim.deep_equal(pred, { 'set!', 'injection.combined' })
+        end, { predicate = true })
```
It's too late at that point. `_get_injection` is invoked after `iter_matches`, but we need to check the existence of combined injection before calling `iter_matches` to compute the `start_line`/`end_line` arguments.

tomtomjhj commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -822,30 +965,47 @@ end
-        self:_log('match from injection query failed for pattern', pattern)
+    local start_line, _, end_line, end_col = root_node:range()
+    if end_col > 0 then
+      end_line = end_line + 1
```
```
languagetree.lua:983:7: value assigned to variable end_col is unused
```
:/

tomtomjhj commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -277,6 +291,10 @@ function LanguageTree:is_valid(exclude_children)
   end
 
   if not exclude_children then
+    -- TODO: what's the proper semantics of is_valid?
+    -- * what doc says: "is_valid() ↔ all trees in for_each_tree are up-to-date"
+    -- * how it's actually implementated and used: "need to process something".
+    --   * if we really want this, then is_valid should take a range (nil ↔ exclude children)
```
I'd appreciate some opinion on this. Some test failures are related to this.

tomtomjhj replied:
relaxed the guarantee provided by `is_valid()`

tomtomjhj commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -623,26 +739,46 @@ function LanguageTree:set_included_regions(new_regions)
+        self._valid = {}
+        for j, _ in pairs(self._regions) do
+          self._valid[j] = true
+        end
+        self._valid[i] = false
+      elseif type(self._valid) == 'table' then
+        self._valid[i] = false
+      end
+      self._regions[i] = region
     end
-    self._trees = {}
-    self:invalidate()
-  else
-    self:_iter_regions(function(i, region)
-      return vim.deep_equal(new_regions[i], region)
-    end)
+
+    touched[i] = true
   end
 
-  self._regions = new_regions
+  -- Discard stale regions. If redrawing, defer it to _finish_batch.
```
It first adds new regions and then discards stale regions. So the behavior of region indices are different from the old version. Some test failures are related to this. Sticking to the old behavior would add some overhead and sometimes just impossible due to batched cleanup.

tomtomjhj replied:
modified the tests to not assume how indices behave

tomtomjhj commented:
note: I'll be offline for the next 3 weeks. I think it's working pretty well, but if anyone finds a problem and is interested in fixing it, please feel free to take over.

lewis6991 commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -131,13 +135,27 @@ function LanguageTree.new(source, lang, opts, parent_lang)
+        break
+      end
+    end
+  end
```
Can this block go into its own function that takes the query as an argument?

```lua
if self._injection_query then
  self._has_combined_injection = query_has_combined(self._injection_query)
end
```

lewis6991 reviewed (changes requested):
This looks good, but I'm not really convinced with the `_batch_*` method here. It makes the languagetree significantly more complicated and is only optimized for the highlighter. Any plugins that need to parse specific buffers with a range aren't going to see any of the benefits.

What I suggest we do is to look at what Helix has done with incremental injections, and see if we can borrow anything from them. https://github.com/helix-editor/helix/blob/0c81ef73e17a3d45cd6240fd5933ad99b3a60d01/helix-core/src/syntax.rs#L1012

Once we've solved all our invalidation bugs, then maybe we can rethink the `parse()` API to better handle multiple ranges due to multiple windows, if adding/removing the languages really is a significant cost. I suspect it isn't common enough for a buffer to be visible in multiple windows at once and thus this isn't something we should be optimizing for.

Another route (and potentially better) would be to make changes to the decoration provider API to better accommodate batch updates, i.e. by calling `parse()` once after all the `on_wins` have been invoked so we can collect the ranges.

clason commented:
> I suspect it is very uncommon for a buffer to be visible in multiple windows at once and thus this isn't something we should be optimizing for.

I don't think it's _that_ uncommon -- having the the definition of a (local) function and its calling site visible at the same time seems like something people do regularly (at least I do). I'd expect this to be more common the longer the file (and thus the more expensive the parsing), too. 

Whether that warrants specific optimization (you'll probably only editing one site) is another question, of course.

lewis6991 commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -263,7 +278,10 @@ end
+---@param exclude_children boolean|nil whether to ignore the validity of children (default `false`).
+--- `is_valid(false)` is sound but not complete, i.e.,
+--- * if it returns `true`, this LanguageTree is actually valid; but
+--- * even if this LanguageTree is actually valid, it may return `false`.
```
This documenting bugs and defining it has a part of the API. 

`is_valid(false)` should always return true if the languagetree is valid.

tomtomjhj replied:
making it complete is possible but that would incur quite a bit of overhead (see note below), and i thought that kinda defeats the purpose of this function (quickly and soundly check necessity of `parse(true)`). i will implement it when i return

tomtomjhj replied:
fixed by running full injection in `is_valid(false)`

https://github.com/neovim/neovim/pull/26827/commits/3d1560bc8101a6e6cc4d641f7726bb1a4d1c951f#diff-70b0f3fa7d6baf9e605400a9f0e10431183df821ec4096e8097eaa5f4d5531baR310-R314

lewis6991 commented:
> I don't think it's that uncommon -- having the the definition of a (local) function and its calling site visible at the same time seems like something people do regularly (at least I do). I'd expect this to be more common the longer the file (and thus the more expensive the parsing), too.
>
> Whether that warrants specific optimization (you'll probably only editing one site) is another question, of course.

Sure, I do this all the time. However, at the very most this is in a split with the buffer in at most 2 windows. Each additional window is exponentially less likely.

tomtomjhj commented:
iiuc helix doesn't do paritial injection: https://github.com/helix-editor/helix/blob/0c81ef73e17a3d45cd6240fd5933ad99b3a60d01/helix-core/src/syntax.rs#L1097
though the idea of constructing lookup table from region to its index on each update is taken from helix.  

the batching is primarily for correcteness of highlights. without it the highlights will ficker when the buffer is shown in multiple windows as the regions created by on_win of a window are discarded by on_win of another window.

i agree that better decoration api for batching is a more natural approach

lewis6991 commented:
> the batching is primarily for correcteness of highlights. without it the highlights will ficker when the buffer is shown in multiple windows as the regions created by on_win of a window are discarded by on_win of another window.

That shouldn't be the root cause of the flicker. Once all the lines for a window have been drawn during a redraw cycle, subsequent windows shouldn't cause flicker in previous windows unless something else is invoking another redraw.

But also, the regions being discarded is just a bug in our highlighter we need to fix. We shouldn't be layering on API methods to work around this.

lewis6991 commented:
> the batching is primarily for correcteness of highlights. without it the highlights will ficker when the buffer is shown in multiple windows as the regions created by on_win of a window are discarded by on_win of another window.

That shouldn't be the root cause of the flicker. Once all the lines for a window have been drawn during a redraw cycle, subsequent windows shouldn't cause flicker in previous windows unless something else is invoking another redraw.

But also, the regions being discarded is just a bug in our highlighter we need to fix. We shouldn't be layering on API methods to work around this.

> iiuc helix doesn't do paritial injection: https://github.com/helix-editor/helix/blob/0c81ef73e17a3d45cd6240fd5933ad99b3a60d01/helix-core/src/syntax.rs#L1097
though the idea of constructing lookup table from region to its index on each update is taken from helix.

The query may not be incremental, but the parsing is.

tomtomjhj commented:
> > the batching is primarily for correcteness of highlights. without it the highlights will ficker when the buffer is shown in multiple windows as the regions created by on_win of a window are discarded by on_win of another window.
> 
> That shouldn't be the root cause of the flicker. Once all the lines for a window have been drawn during a redraw cycle, subsequent windows shouldn't cause flicker in previous windows unless something else is invoking another redraw.

hmm in fact, I can't no longer reproduce this flicker.

So the only valid role of batching is optimization: prevention of discarding regions and adding/removing children.

> But also, the regions being discarded is just a bug in our highlighter we need to fix. We shouldn't be layering on API methods to work around this.

In this PR, `set_included_regions(new_regions)` discards the existing regions (and the corresponding trees) that don't match a region in `new_regions`, where `new_regions` are the regions found by the injection query in the visible range.
I think this is necessary for two reasons:
* to clean up garbage regions: Regions deleted from the source should be deleted from LanguageTree. Also there can be "false un-hits" when matching the old and new regions (see comment in `set_included_regions`).
* to keep number of regions low: LanguageTree would have to manage a large set of regions if the source is big, which adds non-trivial cost to `_edit`.

So there should be some mechanism to discard stale regions.

This PR eagerly removes stale region, which necessitates batching. But it can be avoided if we use less eager method such as assigning some lifespan to each region.

> > iiuc helix doesn't do paritial injection: https://github.com/helix-editor/helix/blob/0c81ef73e17a3d45cd6240fd5933ad99b3a60d01/helix-core/src/syntax.rs#L1097
> > though the idea of constructing lookup table from region to its index on each update is taken from helix.
> 
> Ah true, it's only the parsing that is incremental, but we already do that. However I think their invalidation logic is less buggy than ours, so that's probably what we should fix first.

IIUC Helix is not incremental in the sense that it doesn't care about what's visible on the screen. Also I think their invalidation logic isn't much different from this PR's.

Here's a summary of their approach.

Their `Syntax` struct roughly corresponds to our `LanguageTree`. Like `LanguageTree`, `Syntax` maps an ID to to each `LanguageLayer` (region + tree for that region + various info), but it does not maintain the tree of parsers.

`Syntax::update` is called after each edit.
1.  Apply edits to each layer.
    * update the ranges of affected layers
        * I'm not sure why they do this. This can be delegated to treesitter (which is what nvim does).
    * marks each layers: modified, moved
1.  Construct layer value to layer ID mapping. This is used for checking if a discovered injection region already exists.
1.  Parse layers, starting from the root layer
    1. Edit the tree of the layer.
    1. If the layer was modified by the edit, parse it.
    1. Run the injection query of this layer in the full source.
    1. Check if each injection region matches an existing layer. If so, get its ID, and otherwise, add a new layer.
    1. Repeat the above for each injection layers.
1.  Discard the layers that are not touched by the above process.

A notable difference from nvim is that `Syntax::update` is called after each edit without info about visible range. So Helix is not incremental.

tomtomjhj committed:
perf(treesitter): run injection query only for visible lines

Problem:
Executing injection query on the full source is slow.

Solution:
Execute injection query only on the given range.

Notes
* This is not applicable to languages with combined injection.
* `is_valid(false)` should run full injection to determine if the
  current set of children parsers and their regions are complete. Since
  this can be slow, `parse()` no longer checks this at the beginning.
* Children parsers and regions outside the given range are discarded.

tomtomjhj committed:
perf(treesitter)!: incremental invalidation

Problem:
After an edit that changes the number of injection regions, the
LanguageTree drops all the existing trees. This inefficient because the
injections should be parsed from scratch.

Solution:
When setting included regions, match them with the existing regions so
that they can be reparsed incrementally. This uses a table that maps
region values to their indices. Regions are matched by "similarity",
because some changes of regions cannot be precisely tracked by
`_edit()`.

Breaking change:
The indices of `parser:trees()` behave now differently because existing
regions are reused. So `parser:parse(true)` does not ensure that the
tree table is list-like. Also, when new regions are added manually, they
are first added and then the stale regions are discarded. So the
existing uses of `trees[1]` may break. Use `next(trees())` instead.

tomtomjhj commented:
Since the last comment, I have fixed some bugs, refactored, and added some tests.

Also, I reorganized the commits:
* perf(treesitter)!: incremental invalidation
    * This improves how LanguageTree manages the regions. Specifically, it implements "TODO(lewis6991): inefficient; invalidate trees incrementally".
* perf(treesitter): run injection query only for visible lines
    * This reduces the injection query execution time, which is the most significant optimization in this PR.
* perf(treesitter): batched invalidation
    * This improves region management for the partial injection + multi-window case.

I think the only unresolved questions are:
* Whether the optimization for the multi-window case is worth the complexity (https://github.com/neovim/neovim/pull/26827#issuecomment-1876875040). And similarly, whether incremental invalidation is worth the complexity (new question).
    * It would actually make sense not to include these if the maintainers really want to avoid complexity. The performance improvement from these optimizations would be unnoticeable in most cases. Usually there are not too many visible regions, and each region isn't too big, thus reparsing them from scratch would be fast enough. But I think these optimizations aren't overly complex (after refactoring and commit reorg) and it seems a bit odd not to have them, given that Helix implements incremental invalidation.
* Whether batched invalidation is the best method for optimizing the multi-window case (alternatives: https://github.com/neovim/neovim/pull/26827#pullrequestreview-1803885060)
    * Batched invalidation appears to be the simplest solution. Other solutions would require modifying the decoration provider API, which I'm a bit hesitant to do as I'm not very familiar with the internals of redrawing process.

lewis6991 commented:
Can we remove the batching for now please? I think that over complicates the interface and isn't worth the benefit IMO

clason commented:
If the multi-window optimization can be separated out, I would strongly prefer to keep this out of _this_ PR and investigate it as a follow-up. The base PR is already a good improvement, and this way we can try getting _that_ into 0.10 already.

tomtomjhj commented:
Dropped the batched invalidation commit (https://github.com/tomtomjhj/neovim/commit/fddf77638a8241d24c9250395626b3ad729c495a).

> (Also, some docs seem to be garbled?)

where?

clason commented:
The generated `treesitter.txt` changes.

tomtomjhj commented:
looks fine to me. this was removed because it is no longer true. 
```diff
-    child languages. The result is list-like if
-    • this LanguageTree is the root, in which case the result is empty or a
-      singleton list; or
-    • the root LanguageTree is fully parsed.
+    child languages.
```

clason commented on runtime/doc/treesitter.txt:
```diff
@@ -1321,7 +1321,9 @@ LanguageTree:is_valid({exclude_children})            *LanguageTree:is_valid()*
     Parameters: ~
       • {exclude_children}  (`boolean?`) whether to ignore the validity of
-                            children (default `false`)
+                            children (default `false`). `is_valid(false)` can
+                            be slow because it runs injection on the full
+                            source.
```
And this? I think that belongs to the top-level docs, not the parameter?

tomtomjhj replied:
fixed

tomtomjhj committed:
perf(treesitter): allow parsing multiple ranges

Problem:
Partial injection invalidates regions and children parsers outside the
visible range (passed to `parse`). Invalidating non-matching regions for
each `parse()` is not efficient if multiple windows display different
ranges of the same buffer.

Solution:
Let `parse()` take set of ranges, and invoke `parse()` for all visible
ranges at `on_start`.

tomtomjhj commented:
> [...] maybe we can rethink the parse() API to better handle multiple ranges due to multiple windows, [...]
> 
> Another route (and potentially better) would be to make changes to the decoration provider API to better accommodate batch updates, i.e. by calling parse() once after all the on_wins have been invoked so we can collect the ranges.

Turns out this is actually quite easy to implement. The last commit optimizes the multi-window scenarios by making `parse` take set of ranges and parsing all the visible stuff in decoration provider `on_start` callback (using `line('w0')` and `line('w$')`). No awkward batching stuff.

lewis6991 reviewed (approved):
Looks really good now. Well done!

Just in case this does cause some regressions, maybe better hold off until 0.10 is released, but happy to be overruled on that.

lewis6991 commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -600,6 +654,117 @@ function LanguageTree:_iter_regions(fn)
+---@param region1 Range6[]
+---@param region2 Range6[]
+---@return boolean
+local function region_similar(region1, region2)
```
Needs a description. What is "similar"?

tomtomjhj replied:
added some comment 
<https://github.com/neovim/neovim/compare/058d79cf7464732fc4fa5400994c614b49293c70..7b7736ced849114dcab5f3140e1565452d1ddb3d#diff-70b0f3fa7d6baf9e605400a9f0e10431183df821ec4096e8097eaa5f4d5531baR657-R659>
<https://github.com/neovim/neovim/compare/058d79cf7464732fc4fa5400994c614b49293c70..7b7736ced849114dcab5f3140e1565452d1ddb3d#diff-70b0f3fa7d6baf9e605400a9f0e10431183df821ec4096e8097eaa5f4d5531baR795-R806>

lewis6991 commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -98,6 +113,21 @@ local LanguageTree = {}
 
+---@param injection_query vim.treesitter.Query
+---@return boolean
+local function query_has_combined(injection_query)
```
We may want to move this logic into `vim.treesitter.Query` and have it also calculate some other metadata when loading a query.

tomtomjhj replied:
what about doing this in `Query.new` and storing the result in a field? `query.has_combined`

lewis6991 replied:
That's basically what I'm suggesting. Doesn't need to be done this PR but keep it in mind.

tomtomjhj replied:
done 
<https://github.com/neovim/neovim/compare/7b7736ced849114dcab5f3140e1565452d1ddb3d..1dc295cdb71d171b9f7bbf89a802e42db63225c6#diff-c9baf58a11444ecc08b1a8facc69b0bf9a8033d1b3a2771c6ccefd161e987b79>

clason commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -628,26 +804,45 @@ function LanguageTree:set_included_regions(new_regions)
+--- @param region Range6[]
+--- @return Range6[]
+local function prune_empty_ranges(region)
+  return vim
+    .iter(region)
+    :filter(function(r)
+      return r[3] ~= r[6]
+    end)
+    :totable()
```
Hold that thought; `tbl_filter` is slated for deprecation in favor of -- wait for it -- `vim.iter():filter():totable()`.

lewis6991 replied:
I disagree with that deprecation. We should not view `vim.iter` as a replacement for any `vim.tbl_*` functions. The entire point of `vim.iter` is for chaining operations.

clason replied:
Then please make that case at https://github.com/neovim/neovim/issues/24572; I'm just the messenger.

lewis6991 replied:
Looks like there is no _immediate_ plan: https://github.com/neovim/neovim/issues/24572#issuecomment-2074647896

justinmk replied:
> The entire point of `vim.iter` is for chaining operations.

That rule came from nowhere. The point of vim.iter as described in https://github.com/neovim/neovim/issues/18585 is to provide an interface for iterables.

lewis6991 replied:
That rule came from the author of `vim.iter`, which has been mentioned more than once in chat. Sorry to pull you in @gpanders but am I wrong here?

This is become a noticable problem since contributers have been reaching for `vim.iter` unnecessarily when almost all of the time a regular `pairs`/`ipairs` loop is enough.

gpanders replied:
I agree that we do not need to use vim.iter every single time we need to perform a simple filter operation. Likewise for tbl_filter. I agree with @lewis6991, a regular for loop is often more readable and even faster.

vim.iter provides a generic iterator interface which is designed for working with pipelines. If your pipeline has only a single stage, then it’s just a complicated way of writing a for loop. Lua already has a “builtin” iterator concept so why not leverage that when it makes sense?

tomtomjhj replied:
Replaced with probably more efficient(?) in-place update
<https://github.com/neovim/neovim/compare/058d79cf7464732fc4fa5400994c614b49293c70..7b7736ced849114dcab5f3140e1565452d1ddb3d#diff-70b0f3fa7d6baf9e605400a9f0e10431183df821ec4096e8097eaa5f4d5531baR847-R858>

justinmk replied:
> If your pipeline has only a single stage, then it’s just a complicated way of writing a for loop.

No disagreement there. I just want to make it clear that `vim.iter` is positioned for all kinds of iterables, to avoid the idea that we need table-only or list-only analogs for its functionality.

tomtomjhj committed:
docs: news

tomtomjhj committed:
fix(treesitter): don't use `[1]` and `ipairs` on `trees()`

|LanguageTree:trees()| no longer guarantees that the returned table is
list-like even after a full parse.

NOTE: Actually, they should work fine without modification in these
specific cases.
* dev.lua: For the root tree, the returned table is always `{ tree }`
  unless the private method `LanguageTree:set_included_regions` is used.
* gen_help_html.lua: `trees()` is initially list-like in the very first
  parse. So `ipairs` should be fine if source is not modified.

However, it seems better to "fix" them for consistency.

tomtomjhj commented:
* Added entries in news.txt
* "fixed" some usages of `trees()` that are no longer valid according to the documentation.

lewis6991 commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -585,6 +600,120 @@ function LanguageTree:_iter_regions(fn)
+
+---@param regions_inv table<integer, (Range6[]|integer)[]>
+---@param region Range6[]
+local function regions_inv_remove(regions_inv, region)
```
`_discard_region` is called in a loop which calls this function. Each time you call `table.remove` it causes a full table re-index.

Should this be optimized somehow or do you think it won't have much effect?

tomtomjhj replied:
The bucket is very likely to be small, usually contains a single region. So this should be fine.

clason commented on runtime/doc/news.txt:
```diff
@@ -119,7 +121,9 @@ TERMINAL
 
-• TODO
+• |LanguageTree:parse()| runs injection query only on the provided list of
+  ranges as long as the langauge does not have a combined injection,
+  signifincantly improving |treesitter-highlight| performance.
```
```suggestion
  ranges as long as the language does not have a combined injection,
  significantly improving |treesitter-highlight| performance.
```

lewis6991 commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -649,12 +818,24 @@ function LanguageTree:included_regions()
   for i, _ in pairs(self._trees) do
-    regions[i] = self._trees[i]:included_ranges(true)
+    local region = self._trees[i]:included_ranges(true)
+    prune_empty_ranges(region)
```
This needs explaining. Why is treesitter giving back empty ranges? Is that a bug?

tomtomjhj replied:
This happens when the user edits a region in a way that deletes a range in it.

Consider the following markdown text.  
```
* 1
  2
```
Initially, the markdown_inline parser has a single region `{ { 0, 2, 2, 1, 0, 4 }, { 1, 2, 6, 1, 3, 7 } }`.

If you delete the second line, it becomes `{ { 0, 2, 2, 1, 0, 4 }, { 1, 0, 4, 1, 0, 4 } }`. This means that `tree:edit()` doesn't delete empty ranges itself. 

Maybe this could be considered a bug in treesitter.

lewis6991 replied:
EIther way, this is useful information that should be included in the code.

lewis6991 commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -649,12 +818,24 @@ function LanguageTree:included_regions()
+      regions_inv_insert(regions_inv, i, region)
+    else
+      self._trees[i] = nil
+      if type(self._valid) == 'table' then
```
What about the case when `self._valid` is a boolean?

tomtomjhj replied:
If it's boolean, its value doesn't need to be changed.
* if was `true`: the remaining regions are all valid. so still `true`
* if was `false`:
  * if the removed region is the only region in this language tree, any of `false`/`true` is fine
  * otherwise, remaining regions are invalid. so still `false`.

lewis6991 commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -77,14 +77,24 @@ local TSCallbackNames = {
+---Inverse of `_regions`. start byte of range 1 ↦ { region1, index1, region2, index2, .. }.
+---Used for checking if a new region is already managed by this parser, so that it can be parsed
+---incrementally.
+---@field private _regions_inv table<integer, (Range6[]|integer)[]>?
```
I find this field quite confusing.

Would it be easier to split this into two table, one for the regions and one for the indexes?

lewis6991 commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -77,14 +77,24 @@ local TSCallbackNames = {
 ---taken from _trees. This is mostly a short-lived cache for included_regions()
+---@field private _regions table<integer, Range6[]>?
+---
+---Inverse of `_regions`. start byte of range 1 ↦ { region1, index1, region2, index2, .. }.
```
I don't understand this. Can it be explained more?

tomtomjhj replied:
Copied from the comments I added in the latest force push:

Inverse region table, i.e., a (chaining) hash table from regions to their index in `_region`.
Used for checking if an added region is already managed by this parser, so that it can reuse
the existing tree for incremental parsing.
The hash function is simply `region[1][3]` (the start byte of its first range).
Each bucket has the shape of { region1, index of region1, region2, index of region2, ... }.

lewis6991 commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -585,6 +600,120 @@ function LanguageTree:_iter_regions(fn)
+
+---@param regions_inv table<integer, (Range6[]|integer)[]>
+---@param region Range6[]
+---@return integer?
+---@return boolean? exact
```
Can you add descriptions for these.

lewis6991 reviewed (commented):
Giving this a closer look now and it looks quite complicated. We may want to think about promoting the region stuff into its own abstraction via a class or module.

tomtomjhj commented:
Bug: When scrolling in diff mode, the alignment of diffed windows can be broken.
![image](https://github.com/user-attachments/assets/6bdcf09a-7dec-4d47-8879-fda3b0f7d8e0)

This seems to happen due to the use of `line("w0")` in decoration provider `on_start`. `line("w0")` calls `update_topline`, which might be messing up with the diff alignment. 
https://github.com/neovim/neovim/blob/9788b81d7e53e7ad1ca9db249d86e43426ce56c2/src/nvim/eval.c#L6532-L6537

lewis6991 commented:
Looks unrelated to this change and should probably be reported as a separate issue.

tomtomjhj commented:
Yes. I can reproduce in vim by manually invoking `line("w0")` in diff mode. I'll file an issue there.

justinmk commented:
See https://github.com/neovim/neovim/pull/31631#issuecomment-2557929605

tomtomjhj commented:
Rebased. I'd like to revisit this and introduce some abstraction as lewis requested, but I'll be quite busy until mid-January.

ribru17 commented:
Correct me if I'm wrong, but it looks like some extra code is added to preserve injections in-between combined ranges; i.e. if we have combined injections for comments

```lua
-- inject here

... not here

-- inject here
```

this PR handles that case and the "not here" will have its range parsed since it is betwixt two combined regions. I think it would be good to not worry about this case and simplify the PR (if it does in fact simplify it). Reason being that this behavior is in fact a flaw in neovim; the true tree-sitter documented behavior is to have all "combined" regions treated as one document/tree (but have their ranges still be separated). In neovim we incidentally treat them also as one contiguous range since there is not (yet) a notion of one injected tree with multiple separate ranges. Eventually this combined range functionality will be fixed (removed) and thus I feel like it is best not to complicate this PR by trying to support behavior that shouldn't even exist in the first place

tomtomjhj commented:
@ribru17   hmm i dont quite follow the discussion about desired bavavior of combined injection.
is this what you're referring to? https://github.com/tree-sitter/tree-sitter/issues/3517

besides, i dont recall adding new code related to combined injection other than disabling the partial injection (might be wrong. i'm not looking at the code now)

ribru17 commented:
> is this what you're referring to? [tree-sitter/tree-sitter#3517](https://github.com/tree-sitter/tree-sitter/issues/3517)

yeah, exactly

I guess another way to phrase it is: if we *don't* add the code to disable partial injection for combined injections, what is the result? Is it that the ranges inbetween do not get injected? If so, I think that is not only acceptable but actually better (closer to intended treesitter behavior)

lewis6991 commented:
How so? That's not been my interpretation. Incremental injections is just a case of performance, something treesitter makes no statements about. And I'm not sure what you interpretation really is based off of your example. My understanding is that every combined range is interpreted as a single region, but that region can have gaps in, but from the parsers POV it is just a single range.

ribru17 commented:
Yeah I'm not arguing the performance aspect. This was my understanding:

- Currently injection queries are run through the entire buffer range (slow for huge files)
- Solution in this PR: only run them for ranges of nodes where we mark `@injection.content`
  - But there is a problem! injection.combined gives us a range not only of those nodes, but also all ranges inbetween
  - Thus we add more code to handle the special case of injection.combined, which is to still run injection queries over nodes in between injection.combined nodes

Let me know if any of that is incorrect. But if not I am just saying that we should not worry about that special case because we are doing more work to support a current aspect of injection functionality which is a bug (by treesitter's definition)

ribru17 commented:
Just to give an example to clarify the differences:

Neovim's injection.combined (for comment content(!) only, not comment markers):

```lua
-- print([[ some
-- text
-- here]])
```

in Neovim the comment markers will get the `@string` highlight (bad)

in `tree-sitter highlight`, the comment markers are *not* included in this range (since they are not captured as `@injection.content` but we still have all 3 comments parsed as one singular tree (just without the contiguous range). 

(If it was not one contiguous tree, e.g. `here]])` would be its own document and would be a parse error)

lewis6991 commented:
> in Neovim the comment markers before text and here]]) will get the @string highlight (bad)

I forgot neovim still has that bug, but I think that is beside the point? Even if we correctly excluded the comment markers from the combined region, we would still want to disable incremental injections for combined injections because we need to scan the whole document to get the full combined region. Whether or not we've collected the ranges correctly is a different matter.

ribru17 commented:
Gotcha, that makes sense. I had thought the "partial" aspect of this PR only applied to the query iteration + highlight applications. But yeah you're right if it also includes the actual ranges of what to parse then it would have to say

ribru17 commented on runtime/lua/vim/treesitter/query.lua:
```diff
@@ -88,6 +89,17 @@ function Query.new(lang, ts_query)
+        return vim.deep_equal(directive, { 'set!', 'injection.combined' })
+      end, { predicate = true })
+    then
+      self.has_combined_injection = true
```
I think this logic should be removed, and all special cases regarding combined injections here should be removed also. Per my previous comment. As Lewis pointed out, things are different when it comes to parsing, but querying should only run on the injection ranges, even for combined injections. Combined injections only apply to the tree document content, *not* to the ranges for which highlights should be applied

ribru17 commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -1007,30 +1021,48 @@ end
+    -- If the query doesn't have combined injection, run the query on the given range. Combined
+    -- injection must be run on the full range. Currently there is no simply way to selectively
+    -- match each pattern separately.
+    if range ~= true and not self._injection_query.has_combined_injection then
```
Similarly here, I think this check should be removed. And the other one earlier in this file

ribru17 commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -439,11 +448,6 @@ end
 ---     only the root tree without injections).
 --- @return table<integer, TSTree>
 function LanguageTree:parse(range)
-  if self:is_valid() then
```
Will this removal cause unnecessary reparsing? Is it possible to maybe pass in a range to `is_valid()`, which will check if the languagetree is valid for that range? And maybe then we could also remove the

```lua
      self:_add_injections(true)
      self._injections_processed = true
```

side effect in `is_valid()` (not sure if that would be possible after this change though)

tomtomjhj replied:
See my reply to your comment on `is_valid()` documentation.

ribru17 commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -72,8 +72,12 @@ local TSCallbackNames = {
+---If `is_valid(true) and _injections_processed`, the set of children parsers and their sets of
+---regions are complete wrt. the full source, so that it's not necessary to execute injections.
 ---@field private _injections_processed boolean
+---
```
Was this line unintentional

ribru17 commented on runtime/doc/treesitter.txt:
```diff
@@ -1551,7 +1551,8 @@ LanguageTree:invalidate({reload})                  *LanguageTree:invalidate()*
     Returns whether this LanguageTree is valid, i.e., |LanguageTree:trees()|
     reflects the latest state of the source. If invalid, user should call
-    |LanguageTree:parse()|.
+    |LanguageTree:parse()|. `is_valid(false)` can be slow because it runs
```
I think it would be really good to avoid this side effect in `is_valid()`, if possible

tomtomjhj replied:
This PR's region management strategy only retains the regions that intersect the requested parsing range. But computing `is_valid(false)` requires the knowledge of all possible regions, including the ones that are not currently tracked. This necessitates running injection.

Actually, I would say `is_valid(false)` is kind of unnecessary. The thing you would mostly likely to do after seeing `is_valid(false) == false` is `parse(true)`. The `is_valid(false) == false` guard for `parse(true)` would be useful if it avoids the cost of `parse(true)` in the `is_valid(false) == true` case. But if you run `parse(true)` in the `is_valid(false) == true` case, `parse(true)` will be no-op. So the `is_valid(false) == false` guard isn't really meaningful. This is why I removed `is_valid()` check in `_parse()`.

ribru17 commented on runtime/lua/vim/treesitter/highlighter.lua:
```diff
@@ -383,6 +383,23 @@ function TSHighlighter._on_spell_nav(_, _, buf, srow, _, erow, _)
+  local buf_ranges = {} ---@type table<integer, (Range)[]>
+  for _, win in ipairs(api.nvim_tabpage_list_wins(0)) do
+    local buf = api.nvim_win_get_buf(win)
+    if TSHighlighter.active[buf] then
```
If this is false, will there be a `nil` access a few lines below, where the parse is called?

tomtomjhj replied:
The loop below loops over `buf_ranges` which contains buffers for which `TSHighlighter.active[buf]` is non-nill.

ribru17 commented:
I believe I have a commit that implements the major perf gain of this patch with a small fraction of the diffstat- would you mind taking a look? https://github.com/neovim/neovim/commit/0cd9b0efcd5fbf474e14439a18f87e34240498bb

Note that some of the implementation might be a bit hacky, like with `self._injections_processed`, but I think it is overall sound. Let me know if I am missing something obvious with this :sweat_smile: If not, maybe a smaller diff could help this PR get merged quicker

tomtomjhj commented:
> I believe I have a commit that implements the major perf gain of this patch with a small fraction of the diffstat- would you mind taking a look? https://github.com/neovim/neovim/commit/6454f81550872ef72172d2830a9f4f8f0e6bd63f

I think this is a reasonable fragment. But for the record, it has the problems mentioned in the following two commits of the current PR, though the impact won't be too bad. 
* perf(treesitter)!: incremental invalidation
* perf(treesitter): allow parsing multiple ranges

> like with self._injections_processed

I think it's clearer to use `boolean|Range` (instead of `string`) and compare the values with `vim.deep_equal`.

ribru17 commented:
> I think this is a reasonable fragment. But for the record, it has the problems mentioned in the following two commits of the current PR, though the impact won't be too bad.

That makes sense. Maybe these can be saved for a follow up PR if the impact isn't too bad? I have also noticed like you say that the incremental validation for injection trees isn't too bad since there usually is a very small amount of code parsed by them

> I think it's clearer to use `boolean|Range` (instead of `string`) and compare the values with `vim.deep_equal`.

That sounds much better :+1:

tomtomjhj commented:
> Maybe these can be saved for a follow up PR if the impact isn't too bad? I have also noticed like you say that the incremental validation for injection trees isn't too bad since there usually is a very small amount of code parsed by them

Yes I agree.

ribru17 commented:
Would you prefer to edit this PR, or would you like me to make one? Feel free to use the patch I posted if you like; I also don't mind making a separate change

tomtomjhj commented:
Please make a new PR

clason commented:
Am I correct in thinking that this PR is now superseded, at least in its current form, and that a fresh PR for the additional changes on top of master would be preferable?

tomtomjhj commented:
Yes.
